src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/LessAndGreaterThanCompletionSession.cs

```diff
@@ -57,6 +59,10 @@ private bool PossibleTypeArgument(ITextSnapshot snapshot, SyntaxToken token, Can
             }
 
             var model = document.GetSemanticModelAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+            if (model == null)
```

Comment:
How did we end up with syntax but no semantic model? That seems...odd?

Reply:
In the remote case, we have the document and syntax information, just not semantic since we only have whatever documents have been opened on the guest side.

Reply:
That doesn't actually make sense (and indicates some sort of invariant being broken).  We should *always* be able to get the semantic model for C#/VB documents.  THere should never be a case where we can have a syntax tree and no semantic model.

Reply:
> THere should never be a case where we can have a syntax tree and no semantic model.

I'm open to changing this to checking if the semantic model is [supported ](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/Document.cs,6ece7fd74334e8e0)on the document rather than null but I don't think that would solve the issue that a syntax tree is provided without a semantic model.  The reason it's null is because we don't provide the `ICompilationFactoryService` (the guest doesn't know how to create the compilation)

Assuming we don't want to break the invariant (it is currently being broken) that having a syntax tree means there is also a semantic model, then we have to figure out two problems
1.  How to provide a very incomplete/invalid semantic model.  The guest doesn't have the information needed to create a valid semantic model.
2.  How to then tell features not to use the semantic model when running as a guest.  We don't want to try and use a semantic model that is not correct.

The current path we're going down is experimenting with making remote syntactic calls where possible so we won't have this issue, but until then it seems like there isn't a good way to enforce the invariant.  If you have better ideas, I'm all ears!

Reply:
> The guest doesn't have the information needed to create a valid semantic model.

You can always make a semantic model from a single tree.  Just make a compilation with just that tree in it, and get the semantic model for it.

> The current path we're going down is experimenting with making remote syntactic calls where possible so we won't have this issue, but until then it seems like there isn't a good way to enforce the invariant

I mean, i think that's ok.  But why are you limiting to just syntactic calls?  It seems like remoting semantic calls would be very desirable as well.  that way the user can offload the semantic processing to some other system.

Reply:
> But why are you limiting to just syntactic calls?

Most semantic calls already are remote, so I guess what I really meant is make **everything** remote including syntactic stuff.

Reply:
> You can always make a semantic model from a single tree. Just make a compilation with just that tree in it, and get the semantic model for it.

Good to know that is an option!  I'll play around a bit with it tomorrow, but I'm not sure what side effects making it non-null now will have.  IMHO performing a null check on a nullable (for unfortunate reasons, but still nullable) isn't a bad fix to stop throwing the exception until we finish our experiment.  We already have an item in the cloud backlog to revisit after we've finished testing out other generally syntactic features

Reply:
> Good to know that is an option! I'll play around a bit with it tomorrow,

Sounds good.  Note: i'm extremely wary about what you're doing.  While experiments are great, i do really worry aobut ones that just fundamentally undo invariants Roslyn has literally had for like 10 years now.  

If this does work out (and i know i'm beating a dead horse here), please don't just elevate this experiment into shipping code.  Too many times in the past the C#/VB teams got saddled with hacked-together systems that hten shipped, but then had to be maintained for years.  It's enormously debilitating for a team.  Regardless of the outcome of the experiment, there needs to be a well-thought-through design here that all the stakeholders think is appropriate for this space and which is also fine for the long term health of Roslyn.